bip-0000.mediawiki

```diff
@@ -0,0 +1,335 @@
+** In the case of ''m'' = 0, no label is applied and ''B<sub>m</sub> = B<sub>spend</sub>''
+* The final address is a [https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki Bech32m] encoding of:
+** The human-readable part "sp" for mainnet, "tsp" for testnets (e.g  signet, testnet, regtest)
+** The data-part values:
+*** The character "q", to represent a silent payment address of version 0
+*** The 64 byte concatenation of the receiver's public keys, ''B<sub>m</sub> || B<sub>scan</sub>''
+
+=== Outpoints hash ===
+
+The sender and receiver MUST calculate an outpoints hash for the transaction in the following manner:
+
+* Collect each ''outpoint'' used as an input to the transaction
+* Let ''outpoints = outpoint<sub>0</sub> || … || outpoint<sub>n</sub>'', sorted by txid and vout, ascending order
+* Let ''outpoints_hash = sha256(outpoints)''
+
+=== Inputs For Shared Secret Derivation ===
+
+While any UTXO with known output scripts can be used to fund the transaction, the sender and receiver MUST use inputs from the following list when deriving the shared secret:
+
+* ''P2TR (key path spends)''
+* ''P2WPKH''
+* ''P2SH-P2WPKH''
+* ''P2PKH''
+
+Inputs with conditional branches or multiple public keys (e.g ''CHECKMULTISIG'') are not included as this introduces malleability and would allow a sender to re-sign with a different set of public keys after the silent payment output has been derived. This is not a concern when the sender controls all of the inputs, but is an issue for CoinJoins and other collaborative protocols, where a malicious participant can participate in deriving the silent payment address with one set of keys and then re-broadcast the transaction with signatures for a different set of public keys. P2TR can have hidden conditional branches (script path), but we work around that as described below.
+
+''' P2TR (key path spends)'''
+
+The sender MUST use the tweaked private key corresponding to the taproot output key (i.e. the tweaked private key for a key path spend). If the taproot output key is not available, the output cannot be used for sending a silent payment. The receiver always uses the taproot output key when scanning, regardless of whether the taproot output is being used in a key path spend or a script path spend<ref name="why_always_output_pubkey">''' Why not skip all taproot script path spends? ''' This causes malleability issues for CoinJoins. If the silent payments protocol skipped taproot script path spends, this would allow an attacker to join a CoinJoin round, participate in deriving the silent payment address using the tweaked private key for a key path spend, and then broadcast their own version of the transaction using the script path spend. If the receiver were to only consider key path spends, they would skip the attacker's script path spend input when deriving the shared secret and not be able to find the funds.</ref>.
```

Comment:
Could this be changed to "The sender MUST be able compute `ax * Bscan` on the tweaked private key (`ax`) corresponding to the taproot output key [...]". I believe that BIP327 (MuSig2) users can compute the correct value to be used with silent payments, but this phrase excludes them using it. This would also require changing the computation of `P0` to something like:
```
let S = a0 * B + a1 * B + ... + an * B
let P0 = hash(outpoints_hash * S || 0) * G + B
```

Reply:
> Glad to see this proposal advancing! I think with some minor changes it can be compatible with MuSig2 Taproot Keypaths (and potentially other off chain key aggregation schemes), which make it a powerful way of transacting privately even between users with advanced keys and scripts on Taproot.

Thanks so much for the review! We definitely want to support key aggregation techniques (Musig2 / FROST / etc) on both the sender (inputs) and receiver (having the spend key be an aggregated key). I need to read up on BIP327 to make sure I fully understand your comments but will respond shortly.  I've also responded in line to your other comments.

Reply:
Thanks for the comments. Is the concern that the way it is worded it sounds like the owner of an input is a single person while in reality it could be a group of owners using MuSig (or even FROST)? I do think it is worth specifically mentioning somewhere, but there shouldn't be anything in the current specification that causes issues for such cases. Shared secrets are linear and so in e.g. a MuSig 2-of-2 the two people can both just calculate the shared secret (with their MuSig tweaked key) and add up the result.

If these two people don't trust each other then they also need to provide an discreet log equivalence proof (relatively simple) to prove they calculated the shared secret correctly. This detail probably is important enough to add to the BIP, now that I think about it.

Reply:
My concern was just the precise phrase `MUST use the tweaked private key` which does not exist in an aggregate key situation. So changing the phrasing be explicit about what computation is required ensures that users of complex keys are not incorrectly deterred. I think the approach for MuSig in particular would be something like...

Participants compute and publish:
```
aB_px = d_px * B
```
Coordinator computes:
```
untweaked_agg_pk = keyagg_coef_p1 * A_p1 + keyagg_coef_p2 * A_p2 + ... + keyagg_coef_pn * A_pn
aB_agg = keyagg_coef_p1 * aB_p1 + keyagg_coef_p2 * aB_p2 + ... + keyagg_coef_pn * aB_pn
aB_agg' = pointNegate(aB_agg) if !has_even_y(untweaked_agg_pk) else aB_agg
aB = aB_agg' + tapTweak * B
```
(please don't quote me on this computation, I think it's close, but probably missed at least one detail)

Of course the above is also complicated by x-only internal keys. yay.

Reply:
I've added a footnote and attempted to update the language to indicate key aggregation techniques are supported in https://github.com/bitcoin/bips/pull/1458/commits/d9e5a1b61dd464b808c8ebc45bbe0104ae8e5e14

@brandonblack let me know if you have any suggestions on how to improve it!